crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+        OSSL_HPKE_err;
+        goto err;
+    }
+    *plainlen = plaintextlen;
+    memcpy(plain, plaintext, plaintextlen);
+
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    EVP_CIPHER_free(enc);
+    OPENSSL_free(plaintext);
+    return erv;
+}
+
+/*
+ * @brief do AEAD encryption as per the RFC
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param suite is the ciphersuite
+ * @param key is the secret
+ * @param keylen is the length of the secret
+ * @param iv is the initialisation vector
+ * @param ivlen is the length of the iv
+ * @param aad is the additional authenticated data
+ * @param aadlen is the length of the aad
+ * @param plain is an output
+ * @param plainlen is the length of plain
+ * @param cipher is an output
+ * @param cipherlen input/output, better be big enough on input, exact on output
+ * @return 1 for good otherwise bad
+ */
+static int hpke_aead_enc(OSSL_LIB_CTX *libctx,
+                         OSSL_HPKE_SUITE suite,
+                         unsigned char *key, size_t keylen,
+                         unsigned char *iv, size_t ivlen,
+                         unsigned char *aad, size_t aadlen,
+                         unsigned char *plain, size_t plainlen,
+                         unsigned char *cipher, size_t *cipherlen)
+{
+    int erv = 1;
+    EVP_CIPHER_CTX *ctx = NULL;
+    int len;
+    size_t ciphertextlen;
+    unsigned char *ciphertext = NULL;
+    size_t taglen = 0;
+    uint16_t aead_ind = 0;
+    EVP_CIPHER *enc = NULL;
+    unsigned char tag[16];
+
+    aead_ind = aead_iana2index(suite.aead_id);
+    if (aead_ind == 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    taglen = hpke_aead_tab[aead_ind].taglen;
+    if (taglen != 16) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if ((taglen + plainlen) > *cipherlen) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /*
+     * Allocate this much extra for ciphertext and check the AEAD
+     * doesn't require more - If it does, we'll fail.
+     */
+    ciphertext = OPENSSL_malloc(plainlen + taglen);
+    if (ciphertext == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Create and initialise the context */
+    if (!(ctx = EVP_CIPHER_CTX_new())) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Initialise the encryption operation. */
+    enc = EVP_CIPHER_fetch(libctx, hpke_aead_tab[aead_ind].name, NULL);
+    if (enc == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_EncryptInit_ex(ctx, enc, NULL, NULL, NULL) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    EVP_CIPHER_free(enc);
+    enc = NULL;
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, ivlen, NULL) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Initialise key and IV */
+    if (EVP_EncryptInit_ex(ctx, NULL, NULL, key, iv) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /*
+     * Provide any AAD data. This can be called zero or more times as
+     * required
```

Comment:
But we only call it zero or once, so the comment seems a bit odd.

Reply:
Yeah, left over from somewhere else most likely, so now gone.